iree/hal/metal/metal_kernel_library.mm

```diff
@@ -0,0 +1,135 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/hal/metal/metal_kernel_library.h"
+
+#include "iree/base/memory.h"
+#include "iree/base/status.h"
+#include "iree/base/tracing.h"
+
+namespace iree {
+namespace hal {
+namespace metal {
+
+// static
+StatusOr<ref_ptr<MetalKernelLibrary>> MetalKernelLibrary::Create(
+    id<MTLDevice> device, ExecutableCachingModeBitfield mode,
+    const MetalExecutableDef& metal_executable_def) {
+  IREE_TRACE_SCOPE0("MetalKernelLibrary::Create");
+  if (!metal_executable_def.entry_points() || metal_executable_def.entry_points()->size() == 0) {
+    return InvalidArgumentErrorBuilder(IREE_LOC) << "No entry points defined";
+  }
+  if (!metal_executable_def.shader_sources() ||
+      metal_executable_def.shader_sources()->size() == 0) {
+    return InvalidArgumentErrorBuilder(IREE_LOC) << "No MSL source string present";
+  }
+
+  const auto& entry_points = *metal_executable_def.entry_points();
+  const auto& msl_sources = *metal_executable_def.shader_sources();
+
+  if (entry_points.size() != msl_sources.size()) {
+    return InvalidArgumentErrorBuilder(IREE_LOC)
+           << "Number of entry points and source strings mismatch";
+  }
+
+  // Compile each MSL source string into a MTLLibrary and get the MTLFunction for the entry point to
+  // build the pipeline state object.
+
+  absl::InlinedVector<id<MTLLibrary>, 1> libraries;
+  absl::InlinedVector<id<MTLFunction>, 1> functions;
+  absl::InlinedVector<id<MTLComputePipelineState>, 1> states;
+
+  MTLCompileOptions* msl_compile_options = [MTLCompileOptions new];
+  msl_compile_options.languageVersion = MTLLanguageVersion2_0;
+
+  auto cleanup = MakeCleanup([&]() {
+    for (id<MTLComputePipelineState> state : states) [state release];
+    for (id<MTLFunction> function : functions) [function release];
+    for (id<MTLLibrary> library : libraries) [library release];
+    [msl_compile_options release];
+  });
+
+  for (int i = 0; i < msl_sources.size(); ++i) {
+    @autoreleasepool {
+      NSError* error = nil;
+
+      NSString* shader_source = [NSString stringWithCString:msl_sources[i]->c_str()
+                                                   encoding:[NSString defaultCStringEncoding]];
+      id<MTLLibrary> library = [device newLibraryWithSource:shader_source
+                                                    options:msl_compile_options
+                                                      error:&error];
```

Comment:
This shader compilation is synchronous: https://developer.apple.com/documentation/metal/mtldevice/1433431-newlibrarywithsource. We should profile this for larger modules and consider making parts of this async.

https://developer.apple.com/documentation/metal/mtldevice/1433351-makelibrary looks useful

At least leave a TODO to profile/optimize.

Reply:
yeah, we should be compiling all these offline and embedding the library - is there a reason to shuffle around source files instead? (we should be using metallib -> newLibraryWithData:error)

Reply:
(note a reason may be "TODO" - which is fine - but that would be a useful thing to wire up right early on as it's much more representative - we should be able to avoid async, for example)

Reply:
+1. The reason is indeed TODO. Also it's easier to marshal source code through for initial bring up. I think it's also quite useful in the future for debugging purposes. So my plan is to support both in the future and making the offline compilation and MTLLibary as default. We should certainly prefer offline compilation and avoid the compilation at runtime at all. I've a TODO here: https://github.com/google/iree/blob/main/iree/compiler/Dialect/HAL/Target/MetalSPIRV/MetalSPIRVTarget.cpp#L126. Also added one here.